src/search/parser/abstract_syntax_tree.cc

```diff
@@ -0,0 +1,489 @@
+    DecoratedASTNodePtr decorated_arg = decorate_and_convert(
+        arg, arg_info.type, context);
+    context.pop_layer();
+
+    if (arg_info.bounds.has_bound()) {
+        context.push_layer("handling lower bound");
+        ASTNodePtr min_node = parse_ast_node(arg_info.bounds.min, context);
+        DecoratedASTNodePtr decorated_min_node = decorate_and_convert(
+            *min_node, arg_info.type, context);
+        context.pop_layer();
+
+        context.push_layer("handling upper bound");
+        ASTNodePtr max_node = parse_ast_node(arg_info.bounds.max, context);
+        DecoratedASTNodePtr decorated_max_node = decorate_and_convert(
+            *max_node, arg_info.type, context);
+        context.pop_layer();
+
+        decorated_arg = utils::make_unique_ptr<CheckBoundsNode>(
+            move(decorated_arg), move(decorated_min_node), move(decorated_max_node));
+    }
+    FunctionArgument function_arg(key, move(decorated_arg), arg_info.lazy_construction);
+    arguments.insert({key, move(function_arg)});
+    return true;
+}
+
+void FunctionCallNode::collect_keyword_arguments(
+    const vector<plugins::ArgumentInfo> &argument_infos,
+    DecorateContext &context, CollectedArguments &arguments) const {
+    unordered_map<string, plugins::ArgumentInfo> argument_infos_by_key;
+    for (const plugins::ArgumentInfo &arg_info : argument_infos) {
+        argument_infos_by_key.insert({arg_info.key, arg_info});
+    }
+    for (const auto &key_and_arg : keyword_arguments) {
+        const string &key = key_and_arg.first;
+        const ASTNode &arg = *key_and_arg.second;
+        context.push_layer("Checking the keyword argument '" + key + "'.");
+        if (!argument_infos_by_key.count(key)) {
+            vector<string> valid_keys = get_keys<string>(argument_infos_by_key);
+            ostringstream message;
+            message << "Unknown keyword argument: " << key << endl
+                    << "Valid keyword arguments are: "
+                    << utils::join(valid_keys, ", ") << endl;
+            context.parser_error(message.str());
+        }
+        const plugins::ArgumentInfo &arg_info = argument_infos_by_key.at(key);
+        bool success = collect_argument(arg, arg_info, context, arguments);
+        if (!success) {
+            ABORT("Multiple keyword definitions using the same key '"
+                  + key + "'. This should be impossible here because we "
+                  + "sort by key earlier.");
+        }
+        context.pop_layer();
+    }
+}
+
+
+/* This function has to be called *AFTER* collect_keyword_arguments. */
+void FunctionCallNode::collect_positional_arguments(
+    const vector<plugins::ArgumentInfo> &argument_infos,
+    DecorateContext &context, CollectedArguments &arguments) const {
+    int num_pos_args = positional_arguments.size();
+    int num_kwargs = keyword_arguments.size();
+    if (num_pos_args + num_kwargs > static_cast<int>(argument_infos.size())) {
```

Comment:
Do you have an earlier place in mind? We have  to do it while decorating. It cannot be earlier because decoration is the step where we add the information from the features. Before decoration, `f(x1, x2, x3)` we don't know what function `f` refers to and how many parameters it expects. During decoration of a function-call node, we first look up the feature and what arguments it expects, then collect the arguments. Whenever we collect a positional argument, we have to match it to the correct function argument, so at this point we have to protect against users providing too many arguments. There is also a potential error where someone passes an argument both as a positional and a keyword argument and this could easily happen if the positional argument was not meant for this keyword (i.e., the user thought that there is one more positional argument). This is why we print such an elaborate error message.

Maybe we could swap the order of collecting keyword and positional arguments and then change the way errors are reported. Maybe a simpler way would be to just extract the error reporting part into its own function and call it just before collecting the arguments.